internal/reminder/reminder.go

```diff
@@ -149,70 +137,77 @@ func (r *reminder) Stop() {
 	})
 }
 
-// storeCursorState stores the cursor state to a file
-// Not thread-safe, should be called from a single goroutine
-func (r *reminder) storeCursorState(ctx context.Context) error {
+func (r *reminder) sendReminders(ctx context.Context) []error {
 	logger := zerolog.Ctx(ctx)
-	logger.Debug().Msg("storing cursor state")
-
-	var buf bytes.Buffer
-	enc := gob.NewEncoder(&buf)
 
-	data := map[string]interface{}{
-		"projectListCursor": r.projectListCursor,
-		"repoListCursor":    r.repoListCursor,
+	// Fetch a batch of repositories
+	repos, err := r.getRepositoryBatch(ctx)
+	if err != nil {
+		logger.Error().Err(err).Msg("unable to fetch repositories")
+		return []error{err}
 	}
-	if err := enc.Encode(data); err != nil {
-		return err
+
+	logger.Info().Msgf("created repository batch of size: %d", len(repos))
+
+	// Update the reminder_last_sent for each repository to export as metrics
+	for _, repo := range repos {
+		logger.Debug().Msgf("updating reminder_last_sent for repository: %s", repo.ID)
+		logger.Debug().Msgf("previous reminder_last_sent: %s", repo.ReminderLastSent.Time)
+		err := r.store.UpdateReminderLastSentById(ctx, repo.ID)
+		if err != nil {
+			logger.Error().Err(err).Msgf("unable to update reminder_last_sent for repository: %s", repo.ID)
+			return []error{err}
+		}
+		// TODO: Send the actual reminders
 	}
 
-	return os.WriteFile(r.cfg.CursorFile, buf.Bytes(), 0600)
+	return nil
 }
 
-// restoreCursorState restores the cursor state from a file
-// Not thread-safe, should be called from a single goroutine
-func (r *reminder) restoreCursorState(ctx context.Context) error {
+func (r *reminder) getRepositoryBatch(ctx context.Context) ([]db.Repository, error) {
 	logger := zerolog.Ctx(ctx)
-	logger.Debug().Msg("restoring cursor state")
-
-	if _, err := os.Stat(r.cfg.CursorFile); os.IsNotExist(err) {
-		return nil
-	}
 
-	data, err := os.ReadFile(r.cfg.CursorFile)
+	logger.Debug().Msgf("fetching repositories after cursor: %s", r.repositoryCursor)
+	repos, err := r.store.ListEligibleRepositoriesAfterID(ctx, db.ListEligibleRepositoriesAfterIDParams{
+		ID:                r.repositoryCursor,
+		Limit:             int64(r.cfg.RecurrenceConfig.BatchSize),
+		MinElapsedSeconds: r.cfg.RecurrenceConfig.MinElapsed.Seconds(),
+	})
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	buf := bytes.NewBuffer(data)
-	dec := gob.NewDecoder(buf)
+	r.updateRepositoryCursor(ctx, repos)
 
-	cursorData := make(map[string]interface{})
+	return repos, nil
+}
 
-	if err := dec.Decode(&cursorData); err != nil {
-		return err
-	}
+func (r *reminder) updateRepositoryCursor(ctx context.Context, repos []db.Repository) {
+	logger := zerolog.Ctx(ctx)
 
-	if val, ok := cursorData["projectListCursor"]; ok {
-		if v, ok := val.(string); ok {
-			r.projectListCursor = v
-		} else {
-			return fmt.Errorf("projectListCursor is not a string")
-		}
+	if len(repos) == 0 {
+		r.repositoryCursor = uuid.Nil
```

Comment:
Since we're filtering the repos we get back from the batch in the database, it's possible that we read e.g. 50 repos that _all_ got filtered.  I think you want to check `RepositoryExistsAfterID` here.

Reply:
I missed this branch 😞 . I updated the logic. Hopefully I haven't missed any logical branches this time.